subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/repositories/resolver/AbstractDependenciesMetadataAdapter.java

```diff
@@ -32,23 +32,27 @@
 import java.util.List;
 import java.util.Map;
 
-public class DependenciesMetadataAdapter extends AbstractList<DependencyMetadata> implements DependenciesMetadata {
+abstract public class AbstractDependenciesMetadataAdapter<T extends DependencyMetadata> extends AbstractList<T> implements DependenciesMetadata<T> {
     private final List<org.gradle.internal.component.model.DependencyMetadata> dependenciesMetadata;
-    private final Map<Integer, DependencyMetadata> dependencyMetadataAdapters;
+    protected final Map<Integer, T> dependencyMetadataAdapters;
     private final Instantiator instantiator;
-    private final NotationParser<Object, DependencyMetadata> dependencyNotationParser;
+    private final NotationParser<Object, T> dependencyNotationParser;
 
-    public DependenciesMetadataAdapter(List<org.gradle.internal.component.model.DependencyMetadata> dependenciesMetadata, Instantiator instantiator, NotationParser<Object, DependencyMetadata> dependencyNotationParser) {
+    public AbstractDependenciesMetadataAdapter(List<org.gradle.internal.component.model.DependencyMetadata> dependenciesMetadata, Instantiator instantiator, NotationParser<Object, T> dependencyNotationParser) {
         this.dependenciesMetadata = dependenciesMetadata;
         this.dependencyMetadataAdapters = Maps.newHashMap();
         this.instantiator = instantiator;
         this.dependencyNotationParser = dependencyNotationParser;
     }
 
+    protected abstract Class<? extends T> adapterImplementationType();
```

Comment:
:x: I think it would be valuable to add both the adapter type and the `pending` boolean as constructor arguments and don't require the implementors to do it.

Reply:
:x: Also looking closer at the implementation classes, there doesn't seem to be any specialization if you push that up. There's no specialization in `AbstractDependencyMetadataAdapter` either, which makes me think it should be possible to flatten all this in a single implementation class.

Reply:
The interfaces that are implemented are different. One implements `DependencyConstraintsMetadata` and one `DependenciesMetadata`.

But I very much agree that this feels unnecessary complex. I just have no good idea what's better/best at the moment. Maybe we should unify constraints and dependencies more... Maybe we know better once we did more work on the other constraints related stories. And see more use case.

How about I leave this as it is for now and create a follow up issue (to be addressed before we declare any of this stable). I think it is ok if we do changes to the new incubating interfaces later (even after the 4.5 release) since the DSL itself can probably stay as it is.

Reply:
https://github.com/gradle/gradle/issues/3732

Reply:
:+1: